crates/compiler/build/src/program.rs

```diff
@@ -762,121 +845,59 @@ fn build_loaded_file<'a>(
-        if !preprocessed_host_path.exists() && legacy_host_path.exists() {
-            linking_strategy = LinkingStrategy::Legacy;
-        }
-    }
-
-    // the preprocessed host is stored beside the platform's main.roc
-    let preprocessed_host_path = if linking_strategy == LinkingStrategy::Legacy {
-        if target == Target::Wasm32 {
-            // when compiling a wasm application, we implicitly assume here that the host is in zig
-            // and has a file called "host.zig"
-            platform_main_roc.with_file_name("host.zig")
-        } else {
-            legacy_host_file(target, &platform_main_roc)
-        }
-    } else {
-        platform_main_roc.with_file_name(roc_linker::preprocessed_host_filename(target))
-    };
+    let output_exe_path = get_exe_path(
+        out_path,
+        app_module_path.as_path(),
+        target,
+        linking_strategy,
+        link_type,
+    );
 
-    let output_exe_path = match out_path {
-        Some(path) => {
-            // true iff the path ends with a directory separator,
-            // e.g. '/' on UNIX, '/' or '\\' on Windows
-            let ends_with_sep = {
-                #[cfg(unix)]
-                {
-                    use std::os::unix::ffi::OsStrExt;
+    let dll_stub_symbols =
+        roc_linker::ExposedSymbols::from_exposed_to_host(&loaded.interns, &loaded.exposed_to_host);
 
-                    path.as_os_str().as_bytes().ends_with(&[b'/'])
+    let built_host_opt =
+        // Not sure if this is correct for all calls with LinkType::Dylib...
+        if link_type == LinkType::Dylib || target == Target::Wasm32 {
+            BuiltHostOpt::None
+        } else {
+            let prebuilt_host = determine_built_host_path(&platform_main_roc_path, target, build_host_requested, link_type, linking_strategy, suppress_build_host_warning);
+
+            match prebuilt_host {
+                BuiltHostOpt::None => {
+                    build_and_preprocess_host(
+                        code_gen_options,
+                        dll_stub_symbols,
+                        emit_timings,
+                        linking_strategy,
+                        &platform_main_roc_path,
+                        &output_exe_path,
+                        target,
+                    )
                 }
-
-                #[cfg(windows)]
-                {
-                    use std::os::windows::ffi::OsStrExt;
-
-                    let last = path.as_os_str().encode_wide().last();
-
-                    last == Some(0x002f)// UTF-16 slash
-                        || last == Some(0x005c) // UTF-16 backslash
+                BuiltHostOpt::Surgical(ref surgical_artifacts) => {
+                    // Copy preprocessed host to executable location.
+                    // The surgical linker will modify that copy in-place.
+                    std::fs::copy(&surgical_artifacts.preprocessed_host, output_exe_path.as_path()).unwrap();
```

Comment:
We also seemingly do this on line 1334 of the same file. Do we need both of these?